### ISSUE TITLE
[Error] WebSockets Request was expected, also faced a CORS issue

### ISSUE THREAD

**Issue by kevinadhiguna** [gitpod-io/gitpod#3831]
### Bug description

Unable to see HTML file and faced a CORS issue.

### Steps to reproduce

1. Create `index.html` and `index.ts` file.
2. Compile `index.ts` and produce an `index.js` file.
3. Create an `<h1>Hello</h1>` and connect it to `index.js`.
4. Open a new tab to access `<PORT>-<Gitpod Workspace URL>`.
5. See a text saying : Web Sockets Request was expected.

### Expected behavior

Display an HTML file and run Javascript in console (not facing a CORS issue)

**Comment by akosyakov** [comment]
Are you sure that you are using the correct port? The server tells you that it exposes web socket connection, not pure http.

**Comment by akosyakov** [comment]
Could you share the repo please? From the description is not clear which server you are running.

**Comment by kevinadhiguna** [comment]
> Could you share the repo please? From the description is not clear which server you are running.

Sure, here is the repo : https://github.com/kevinadhiguna/typescript

**Comment by kevinadhiguna** [comment]
I just ran it, it said it runs on port 42139 : 

![image](https://user-images.githubusercontent.com/43397636/114016641-42036580-9895-11eb-9ace-8211fa2a4834.png)

Then I clicked the Open Browser icon : 

![Screenshot from 2021-04-08 18-08-16](https://user-images.githubusercontent.com/43397636/114016865-8262e380-9895-11eb-8ba7-9e6642638a51.png)

Again, it showed me `WebSockets request was expected ` : 

![image](https://user-images.githubusercontent.com/43397636/114017058-bd651700-9895-11eb-8fd3-92e069306113.png)

**Comment by akosyakov** [comment]
This port is a debug port for Gitpod Code Server. It does not have anythign to do with index.html

Are you trying to preview index.html?

**Comment by kevinadhiguna** [comment]
> This port is a debug port for Gitpod Code Server. It does not have anythign to do with index.html
> 
> Are you trying to preview index.html?

Yes, I am. Sorry I am confused how to preview the HTML file. The docs in Gitpod shows how to do it in Theia but I am a bit confused as I usually drag ad drop an HTML file in local machine to preview it. Do you mind telling me how to preview HTML in a new tab from Gitpod ?

**Comment by akosyakov** [comment]
Hm, VS Code does not support this feature by default, but you can install additional extensions like: https://open-vsx.org/extension/ritwickdey/LiveServer Could you give it a try?

**Comment by kevinadhiguna** [comment]
Sure, I just installed the Live Server extension in VScode in Gitpod.

Next, I right-clicked `index.html` and opened it with Live Server : 

![Screenshot from 2021-04-08 18-36-37](https://user-images.githubusercontent.com/43397636/114020860-38303100-989a-11eb-8bf8-8e899bf6e520.png)

The result was : 

![image](https://user-images.githubusercontent.com/43397636/114020605-fb643a00-9899-11eb-8fbb-f8819faaedb4.png)

A couple of questions please : 
- HTTP 404 indicates that `favicon.ico` was not found, or another error ?
- Why Firefox could not establish connection to the server ? Is there anything wrong ?

**Comment by akosyakov** [comment]
> HTTP 404 indicates that favicon.ico was not found, or another error ?

correct 

> Why Firefox could not establish connection to the server ? Is there anything wrong ?

that's probably related to how we handle ws connection

**Comment by akosyakov** [comment]
I tried in both Chrome Version 89.0.4389.114 (Official Build) (x86_64) and FireFox 87.0 (64-bit) and it worked well. I don't see any errors in FF:
<img width="1408" alt="Screenshot 2021-04-08 at 13 58 22" src="https://user-images.githubusercontent.com/3082655/114022897-892c2f00-9872-11eb-8b80-607c7dafc6a1.png">

Are you sure that you are using in latest FF?

**Comment by kevinadhiguna** [comment]
> > HTTP 404 indicates that favicon.ico was not found, or another error ?
> 
> correct
> 
> > Why Firefox could not establish connection to the server ? Is there anything wrong ?
> 
> that's probably related to how we handle ws connection

Thank you very much, that was so helpful.

**Comment by kevinadhiguna** [comment]
Yes I am using the latest Firefox version 87.0 (64-bit)

![image](https://user-images.githubusercontent.com/43397636/114036960-b4327500-98aa-11eb-9bad-6e373eb14f2b.png)

I once stopped the workspace and the error vanished. However, if I start it again, somehow it appears again in the console..

**Comment by kevinadhiguna** [comment]
Anyway, the Firefox unable to establish connection is not a major issue here IMO. Many thanks for the prompt reply, it helped a beginner like me a lot !

**Comment by akosyakov** [comment]
please upvote here to implement live preview in VS Code: https://github.com/microsoft/vscode/issues/12080 You can also follow this thread for other solutions (VS Code extensions).

**Comment by kevinadhiguna** [comment]
> please upvote here to implement live preview in VS Code: [microsoft/vscode#12080](https://github.com/microsoft/vscode/issues/12080) You can also follow this thread for other solutions (VS Code extensions).

Upvoted, thanks

**Comment by yandeu** [comment]
> A couple of questions please :
> 
> * HTTP 404 indicates that `favicon.ico` was not found, or another error ?
> * Why Firefox could not establish connection to the server ? Is there anything wrong ?

Would it be possible for you to test this with [Five Server](https://marketplace.visualstudio.com/items?itemName=yandeu.five-server)?

**Comment by kevinadhiguna** [comment]
> > A couple of questions please :
> > 
> > * HTTP 404 indicates that `favicon.ico` was not found, or another error ?
> > * Why Firefox could not establish connection to the server ? Is there anything wrong ?
> 
> Would it be possible for you to test this with [Five Server](https://marketplace.visualstudio.com/items?itemName=yandeu.five-server)?

How may I install Five Server in Gitpod ? @yandeu 

![image](https://user-images.githubusercontent.com/43397636/114630241-58218380-9ce4-11eb-8860-0deb0deafa62.png)

**Comment by yandeu** [comment]
> How may I install Five Server in Gitpod ?

I have just added it.

---

Also, I found a bug in Live Server and Five Server that prevents auto-reload on an index page. If you want to use auto-reload, access your index file with `/index.html`.

But everything else (instant update, highlight, auto navigation) works fine.

I will publish an update to Five Server tomorrow.

**Comment by kevinadhiguna** [comment]
Thank you @yandeu , I was able to install Five Server extension.

![image](https://user-images.githubusercontent.com/43397636/114671011-d1919400-9d2d-11eb-9ff3-8c7c9dae259e.png)

However, I encountered some issues regarding connection.
This is how it looks in Firefox :  

![image](https://user-images.githubusercontent.com/43397636/114670410-27197100-9d2d-11eb-935b-f6e0212eacf4.png)

Terminal in Gitpod in FIrefox : 

![image](https://user-images.githubusercontent.com/43397636/114671644-81670180-9d2e-11eb-9a7e-21a2d9f9f65f.png)

This is how it looks in Chrome : 

![image](https://user-images.githubusercontent.com/43397636/114670216-f76a6900-9d2c-11eb-8f13-795d52090a57.png)

Terminal in Gitpod in Chrome : 

![image](https://user-images.githubusercontent.com/43397636/114671707-9479d180-9d2e-11eb-84a9-c84a3e2460a7.png)

This `HTTP 404 indicates that favicon.ico was not found, or another error ?` was replaced with Five Server's favicon.ico, wasn't it ? As it no longer occurred.
Meanwhile, do you have a clue on this issue :  `Why Firefox could not establish connection to the server ? Is there anything wrong ?` 

**Environment** 
Firefox : Version 87.0 (64-bit)
Chrome : Version 89.0.4389.114 (Official Build) (64-bit)
OS : Ubuntu Focal Fossa 20.04.2 LTS

**Comment by kevinadhiguna** [comment]
In firefox, I had to click refresh button to get rid of the error : 

![image](https://user-images.githubusercontent.com/43397636/114672315-37325000-9d2f-11eb-9a1c-c71af48e60b8.png)

Nevertheless, the `Firefox can't establish connection to the server .. ` error happened again after a few seconds.

The same goes to Chrome : 
![image](https://user-images.githubusercontent.com/43397636/114673110-0d2d5d80-9d30-11eb-893c-7afd35b4da4e.png)

**Comment by yandeu** [comment]
> Five Server's favicon.ico, wasn't it ?

Yes 😄  (see https://github.com/yandeu/five-server/commit/5564c401883daeafbb50f0b5d67181a69f856863)

> Why Firefox could not establish connection to the server

Yes, I will try to fix that today. In the meanwhile try to rename your `index.html` to something else like `website.html`.

---

Also, there is a small bug when reloading css. Will fix that as well 👍

**Comment by yandeu** [comment]
@kevinadhiguna I have fixed everything. Just had no time to release it today :/

**Comment by yandeu** [comment]
**Five Server v0.0.25** is now available on gitpod and it works very very well :)

**Comment by kevinadhiguna** [comment]
Cool @yandeu , it worked like magic ! Thanks

![image](https://user-images.githubusercontent.com/43397636/115103831-0fd1c200-9f7f-11eb-9045-9f1663e36af7.png)

**Comment by yandeu** [comment]
You're welcome 🙂

Also, there are new cool features in development.

Currently working on a sort of "Hot Module Replacement" feature. Means, instant update while keeping the state (DOM)  previously manipulated by JavaScript 👍🏻

**Comment by akosyakov** [comment]
@yandeu Does it support preview of PDF files too? In Theia we used to have special handling for pdfs: https://github.com/eclipse-theia/theia/blob/96930a2edcc4c706d41ab8486500de7979d01283/packages/mini-browser/src/browser/mini-browser-content.ts#L581-L593 Maybe you can get inspired from that and add PDf support for Five Server too? We have such questions in the community: https://community.gitpod.io/t/latex-pdf-viewer/3365

**Comment by yandeu** [comment]
I have not implemented anything special for PDFs yet. I guess if you try to preview it, it will display "File is too big for a preview"? I haven't tried it yet. Have you tried it? What does it show?

Five Server allows for middlewares, similar to an express app. So I guess it would be possible to embed a PDF preview.

Feel free to open a new Feature Request Issue on yandeu/five-server (not yandeu/five-server-vscode).

**Comment by akosyakov** [comment]
@yandeu here you go https://github.com/yandeu/five-server-vscode/issues/1

**Comment by akosyakov** [comment]
@yandeu I tried the latest extension and it gives me: 
<img width="1904" alt="Screenshot 2021-04-23 at 08 44 50" src="https://user-images.githubusercontent.com/3082655/115830187-3a54cc80-a410-11eb-9a6a-b2c788f1d89e.png">

From this repo: https://github.com/joejcollins/latex Maybe some CSP is misconfigured for iframe in remote context, not sure?

**Comment by yandeu** [comment]
Yes, I think the CSP policy of the WebView is blocking iFrames.
But it works fine if you try to preview it in the browser :)

**Comment by akosyakov** [comment]
> Yes, I think the CSP policy of the WebView is blocking iFrames.

It is strange though, `frame-src` is set to `*` for the preview extension. Maybe some attributes or policies are missing on level of five server iframe?

**Comment by mfouesneau** [comment]
I'm getting the same issue for the PDF viewer. Any updates?  Could there be information if this issue shows up with vs and not theia maybe?

**Comment by yandeu** [comment]
@mfouesneau @akosyakov Please help to debug the PDF preview feature on https://github.com/yandeu/preview-pdf-test

**Comment by mfouesneau** [comment]
Sure. How can we provide help?

**Comment by yandeu** [comment]
It looks like the "[Simple Browser](https://github.com/microsoft/vscode/blob/main/extensions/simple-browser/src/simpleBrowserView.ts)" preview does not support loading PDFs.

Therefore, Five Server will only show PDF preview in the browser.

I tried [Office Viewer](https://open-vsx.org/extension/cweijan/vscode-office) which seems to show a preview for PDFs.
